### PR TITLE
Fix "request" attribute deprecation on "knp_menu.voter"

### PR DESCRIPTION
**Pull request by TorbenLundsgaard**
I am targeting this branch, because this is BC.

## Changelog

```markdown
### Fixed
- `request` attribute deprecation on `knp_menu.voter`
```

## Subject

Using the "request" attribute of the "knp_menu.voter" tag is deprecated since KnpMenuBundle version 2.2. The RequestStack should be injected in the voter instead.

### CODE REVIEW

**Comment by greg0ire** [comment]
This breaks the test suite for some reason.

File: src/Resources/config/menu.xml [hunk 1]
```diff
@@ -18,7 +18,8 @@
             <tag name="knp_menu.provider"/>
         </service>
         <service id="sonata.admin.menu.matcher.voter.admin" class="Sonata\AdminBundle\Menu\Matcher\Voter\AdminVoter" public="true">
-            <tag name="knp_menu.voter" request="true"/>
+            <argument type="service" id="request_stack" />
```
Comment (greg0ire) [inline src/Resources/config/menu.xml]: There is no constructor argument for this: https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/Menu/Matcher/Voter/AdminVoter.php

File: src/Menu/Matcher/Voter/AdminVoter.php [hunk 2]
```diff
@@ -23,18 +23,32 @@
...
+     * @var RequestStack
+     */
+    private $requestStack;
+
     /**
      * @var Request
      */
     private $request = null;
 
+    public function __construct($requestStack = null)
```
Comment (core23) [inline src/Menu/Matcher/Voter/AdminVoter.php]: Please add the parameter type

Comment (core23) [reply src/Menu/Matcher/Voter/AdminVoter.php]: Also PHPDoc is missing

File: src/Menu/Matcher/Voter/AdminVoter.php [hunk 3]
```diff
@@ -23,18 +23,32 @@
...
     /**
+     * @deprecated since version 3.29. Pass a RequestStack to the constructor instead.
+     *
      * @param Request $request
      *
      * @return $this
      */
     public function setRequest($request)
     {
+        @trigger_error(sprintf('The %s() method is deprecated since version 3.29. Pass a RequestStack in the constructor instead.', __METHOD__), E_USER_DEPRECATED);
```
Comment (core23) [inline src/Menu/Matcher/Voter/AdminVoter.php]: Please don't use the exact version, use `3.x` instead

File: src/Menu/Matcher/Voter/AdminVoter.php [hunk 4]
```diff
@@ -23,18 +24,37 @@
...
      * @var Request
      */
     private $request = null;
 
     /**
+     * AdminVoter constructor.
+     *
+     * @param RequestStack $requestStack
+     */
+    public function __construct($requestStack = null)
```
Comment (core23) [inline src/Menu/Matcher/Voter/AdminVoter.php]:
```diff
- public function __construct($requestStack = null)
+ public function __construct(RequestStack $requestStack = null)
```

**Comment by TorbenLundsgaard** [comment]
As you have probably noticed I'm new to Symfony/Sonata and I'm not sure what to do from here. I'm guessing that the test needs to be changed, but I don't know what to do: https://github.com/sonata-project/SonataAdminBundle/blob/3.x/tests/Menu/Matcher/Voter/AdminVoterTest.php#L42

**Comment by greg0ire** [comment]
@TorbenLundsgaard you have to copy/paste it I guess, and rename it to `createVoterLegacy`, and tag that version with `@group legacy`, and have the other version use the request stack.

**Comment by greg0ire** [comment]
Reviewers: please check the SYMFONY_DEPRECATIONS_HELPER job before merging this.

File: src/Menu/Matcher/Voter/AdminVoter.php [hunk 5]
```diff
@@ -23,18 +24,37 @@
...
+    /**
+     * @deprecated since version 3.x. Pass a RequestStack to the constructor instead.
+     *
      * @param Request $request
      *
      * @return $this
      */
     public function setRequest($request)
     {
+        @trigger_error(sprintf('The %s() method is deprecated since version 3.x. Pass a RequestStack in the constructor instead.', __METHOD__), E_USER_DEPRECATED);
```
Comment (OskarStark) [inline src/Menu/Matcher/Voter/AdminVoter.php]: can you please linebreak this and use a FQCN for RequestStack class?

File: src/Menu/Matcher/Voter/AdminVoter.php [hunk 6]
```diff
@@ -23,18 +24,37 @@
...
+     */
+    private $requestStack;
+
     /**
      * @var Request
      */
     private $request = null;
 
     /**
+     * AdminVoter constructor.
```
Comment (OskarStark) [inline src/Menu/Matcher/Voter/AdminVoter.php]: plaase remove this comment

**Comment by ossinkine** [comment]
Are there any updates on this?

File: src/Menu/Matcher/Voter/AdminVoter.php [hunk 7]
```diff
@@ -47,11 +72,17 @@ public function matchItem(ItemInterface $item)
     {
         $admin = $item->getExtra('admin');
 
+        if (null !== $this->requestStack) {
+            $request = $this->requestStack->getMasterRequest();
```
Comment (ossinkine) [inline src/Menu/Matcher/Voter/AdminVoter.php]: Master or current request should be used here?

Comment (enleur) [reply src/Menu/Matcher/Voter/AdminVoter.php]: https://github.com/KnpLabs/KnpMenu/blob/master/src/Knp/Menu/Matcher/Voter/RouteVoter.php#L59 i guess it should be master request

**Comment by OskarStark** [comment]
Closing in favor of #4869

**Pull request closed without merging**